src/dart/ast.rs

```diff
@@ -0,0 +1,479 @@
+use enum_primitive::FromPrimitive;
+use std::iter;
+use syntax::symbol::Symbol;
+
+#[derive(Debug)]
+pub struct Library {
+    pub name: Option<LibraryName>,
+    pub items: Vec<Item>,
+}
+
+#[derive(Debug)]
+pub struct Combinator {
+    pub hide: bool,
+    pub idents: Vec<Symbol>,
+}
+
+#[derive(Debug)]
+pub struct LibraryName {
+    pub metadata: Metadata,
+    pub path: Vec<Symbol>,
+}
+
+#[derive(Debug)]
+pub struct ImportSpecification {
+    pub uri: StringLiteral,
+    pub deferred: bool,
+    pub as_ident: Option<Symbol>,
+    pub comb: Vec<Combinator>,
+}
+
+#[derive(Debug)]
+pub enum Item {
+    Import(Metadata, ImportSpecification),
+    Export(Metadata, StringLiteral, Vec<Combinator>),
+    Part {
+        metadata: Metadata,
+        uri: StringLiteral,
+    },
+    PartOf {
+        metadata: Metadata,
+        idents: Vec<Symbol>,
+    },
+    Class {
+        metadata: Metadata,
+        abstract_: bool,
+        name: Symbol,
+        params: Vec<TypeParameter>,
+        superclass: Option<Box<Type>>,
+        mixins: Vec<Box<Type>>,
+        interfaces: Vec<Box<Type>>,
+        members: Vec<ClassMember>,
+    },
+    MixinClass {
+        metadata: Metadata,
+        abstract_: bool,
+        name: Symbol,
+        params: Vec<TypeParameter>,
+        mixins: Vec<Box<Type>>,
+        interfaces: Vec<Box<Type>>,
+    },
+    Enum {
+        metadata: Metadata,
+        name: Symbol,
+        values: Vec<Symbol>,
+    },
+    TypeAlias {
+        metadata: Metadata,
+        name: Symbol,
+        params: Vec<TypeParameter>,
+        ty: Box<Type>,
+    },
+    Function(Function),
+    Global(VarType, Vec<NameAndInitializer>),
+}
+
+#[derive(Debug)]
+pub struct Function {
+    pub name: FunctionName,
+    pub params: Vec<TypeParameter>,
+    pub signature: FnSig,
+    pub body: Option<FunctionBody>,
+}
+
+#[derive(Debug)]
+pub enum ClassMember {
+    Constructor {
+        metadata: Metadata,
+        method_qualifiers: Vec<MethodQualifiers>,
+        name: Option<Symbol>,
+        args: FnArgs,
+        initializers: Vec<ConstructorInitializer>,
+        function_body: Option<FunctionBody>,
+    },
+    Method(Metadata, Vec<MethodQualifiers>, Function),
+    Fields {
+        metadata: Metadata,
+        static_: bool,
+        var_type: VarType,
+        initializers: Vec<NameAndInitializer>,
+    },
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum MethodQualifiers {
+    External,
+    Static,
+    Const,
+    Final,
+    Factory,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum FunctionName {
+    Regular(Symbol),
+    Getter(Symbol),
+    Setter(Symbol),
+    Operator(OverloadedOp),
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum OverloadedOp {
+    Tilde,
+    Array,
+    ArrayAssign,
+    Bool(BoolBinOp),
+    Value(ValueBinOp),
+}
+
+#[derive(Debug)]
+pub enum ConstructorInitializer {
+    Redirect(Box<Type>, Option<Symbol>),
```

Comment:
This is actually wrong, my bad, it should be a different variant in `ClassMember`.